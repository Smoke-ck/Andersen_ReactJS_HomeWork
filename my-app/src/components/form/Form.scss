.form {
    &__input {
        margin-bottom: 5px;
        display: block;
        width: 200px;
        height: 10px;
        padding: 10px;
        font-family: inherit;
        font-weight: 400;
        line-height: 1.5;
        color: #212529;
        background-color: #fff;
        border: 1px solid #bdbdbd;
        border-radius: 10px;
        outline: none;
        margin-top: 5px;
        overflow-wrap: break-word;
        &:hover {
            box-shadow:0 0 5px rgba(0,0,0,1);
        }
    }
    &__button {
        background-color:var(--primaryButtonColor);
        border: none;
        border-radius: 5px;
        color: white;
        padding: 5px 20px;
        text-align: center;
        font-size: 16px;
        margin: 4px 2px;
        opacity: 0.6;
        transition: 0.3s;
        display: inline-block;
        text-decoration: none;
        cursor: pointer;
        &:hover {
            opacity: 1;
            color:gold;
        }
    }
}
.errorMsg {
    color:#c30000;
}