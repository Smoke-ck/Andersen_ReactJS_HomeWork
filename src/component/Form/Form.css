.container {
    display: flex;
    flex-direction: column;
}
.formTittle {
    text-align: center;
    font-size: 30px;
}
.errorMsg {
    color: red;
    margin-bottom: 10px;
}
.button {
    vertical-align: top;
    padding: 7px 20px;
    font-family: inherit;
    font-size: 15px;
    text-align: center;
    text-decoration: none;
    width: 150px;
    outline: none;
    border-radius: 10px;
    margin-bottom: 10px;
    border: 1px solid #aaaaaa;
}
.formButton {
    display: flex;
    justify-content: space-between;
}
.button:hover {
    box-shadow:0 0 4px black;
}
.button--ok {
    margin-right: 10px;
    color: green;
}
.button--cancel {
    color:red;
}