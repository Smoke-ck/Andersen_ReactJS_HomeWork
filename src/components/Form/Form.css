.containerForm {
    display: flex;
    flex-direction: column;
}
.formTittle {
    text-align: center;
    font-size: 20px;
}
.errorMsg {
    color: red;
    margin-bottom: 10px;
}
.button {
    cursor: pointer;
    vertical-align: top;
    padding: 7px 20px;
    font-family: inherit;
    font-size: 12px;
    text-align: center;
    text-decoration: none;
    width: 100px;
    outline: none;
    border-radius: 10px;
    margin-bottom: 10px;
    border: 1px solid #aaaaaa;
}
.button:active {
    background-color: white;
}
.formButton {
    display: flex;
    justify-content: space-between;
}
.button:hover {
    box-shadow:0 0 4px black;
}
.button--ok {
    margin-right: 10px;
    color: green;
}
.button--cancel {
    color:red;
}