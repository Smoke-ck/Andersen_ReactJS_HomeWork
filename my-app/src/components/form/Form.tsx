import { FC, useState, } from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { IToDos } from '../../api';
import { createTodo } from '../../store/actions/todos';
import "./Form.scss";

type ITodoForm = {
    createTodo: (title:string) => Promise <IToDos>
}

const Form: FC<ITodoForm> = ({ createTodo }) => {

    const [title, setTitle] = useState('');

    function onValueChange(e: React.ChangeEvent<HTMLInputElement>) {
        setTitle(e.target.value)
    }

    function onFormSubmit(e: React.FormEvent) {
        e.preventDefault();
        createTodo(title.trim());
        setTitle('')
    }

    return (
        <form id="addTaskForm" onSubmit={onFormSubmit}>
            <div className="form">
                <input
                    className="form__input"
                    type="text"
                    id="taskNameInput"
                    value={title}
                    onChange={onValueChange}
                />
                <div className="two columns"> {title.length > 160
                    ? <span className="errorMsg">Превышен лимит текста задачи на {title.length - 160} символов</span>
                    : <input type="submit" className="form__button" value="Добавить" />}
                </div>
            </div>
        </form>
    );
}

const mapDispatchToProps = {
    createTodo
};

export default connect(null, mapDispatchToProps)(Form);