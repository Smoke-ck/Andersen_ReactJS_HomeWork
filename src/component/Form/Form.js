import './Form.css';
import React, { Component } from 'react'
import CustomInput from './CustomInput/CustomInput';
import CustomTextarea from './CustomTextarea/CustomTextarea';

export default class Form extends Component {

    state = {
        newPersonData:
        {
            name: '',
            surname: '',
            phone: '',
            date: '',
            web: '',
            about: '',
            stack: '',
            project: '',
        },
        errors: {},
    };

    handleSubmit = (e) => {
        e.preventDefault();
        if (this.validateForm()) {
            this.props.callbackStateFromForm(this.state.newPersonData)
            this.resetForm();
            this.props.hideForm();
        }

    };

    validateForm() {
        let Data = this.state.newPersonData;
        let errors = {};
        let formIsValid = true;
        let { name, surname, phone, date, web, about, stack, project } = Data 
        name = name.trim();
        if (!name) {
            formIsValid = false;
            errors.name = "Поле пустое"
        } else if ((/[0-9]/).test(name)) {
            formIsValid = false;
            errors.name = "Имя: не может содержать цыфры"
        } else if (!(/[A-Z-А-Я-Ё]/).test(name) || name[0] !== name[0].toUpperCase()) {
            formIsValid = false;
            errors.name = "Имя должно начинаться с большой буквы"
        } 

        surname = surname.trim();
        if (!surname) {
            formIsValid = false;
            errors.surname = "Поле пустое"
        } else if ((/[0-9]/).test(surname)) {
            formIsValid = false;
            errors.surname = "Фамилия: не может содержать цыфры"
        } else if (!(/[A-Z-А-Я-Ё]/).test(surname) || surname[0] !== surname[0].toUpperCase()) {
            formIsValid = false;
            errors.surname = "Фамилия должна начинаться с большой буквы,"
        }

        phone = phone.trim()
        if (!phone) {
            formIsValid = false;
            errors.phone = "Поле пустое"
        } else if (!((/[0-9]{1}-[0-9]{4}-[0-9]{2}-[0-9]{2}/).test(phone)) || phone.length > 12) {
            formIsValid = false;
            errors.phone = "Не коректный номер,формат: X-XXXX-XX-XX"
        }

        if (!web) {
            formIsValid = false;
            errors.web = "Поле пустое "
        } else if (!web.trim().startsWith('https://')) {
            formIsValid = false;
            errors.web = "Имя сайта должно начинаться с 'https://'"
        }
        if (!date) {
            formIsValid = false;
            errors.date = "Поле пустое"
        }

        about = about.trim();
        if (!about) {
            formIsValid = false;
            errors.about = "Поле пустое"
        } else if (about.length > 600) { formIsValid = false; 
        } 

        if (!stack) {
            formIsValid = false;
            errors.stack = "Поле пустое"
        } else if (stack.length > 600) { formIsValid = false; }

        if (!project) {
            formIsValid = false;
            errors.project = "Поле пустое"
        } else if (project.length > 600) { formIsValid = false; }
        this.setState({
            errors: errors
        });
        return formIsValid;
    }

    onChange = (e) => {
        this.setState({
            newPersonData: {
                ...this.state.newPersonData,
                [e.target.name]: e.target.value
            }
        });
    }

    resetForm = () => {
        this.setState({
            newPersonData: {
                name: '',
                surname: '',
                phone: '',
                date: '',
                web: '',
                about: '',
                stack: '',
                project: '',
            },
            errors: '',
        });
    }

    render() {
        return (<>

            <form className="container" onSubmit={this.handleSubmit}>
                <p className="formTittle">Создание анкеты</p>
                <CustomInput
                    type="text"
                    name='name'
                    onChange={this.onChange}
                    state={this.state}
                    label="Имя:"
                />
                <CustomInput
                    type="text"
                    name='surname'
                    onChange={this.onChange}
                    state={this.state}
                    label="Фамилия:"
                />
                <CustomInput
                    type='tel'
                    name='phone'
                    onChange={this.onChange}
                    state={this.state}
                    label="Телефон:"
                />
                <CustomInput
                    type="date"
                    name='date'
                    onChange={this.onChange}
                    state={this.state}
                    label="Дата рождения:"
                />
                <CustomInput
                    type="text"
                    name='web'
                    onChange={this.onChange}
                    state={this.state}
                    label="Веб сайт:"
                />
                <CustomTextarea
                    name='about'
                    onChange={this.onChange}
                    state={this.state}
                    label="  О себе:"
                />
                <div className="errorMsg">{this.state.errors.about}</div>
                <CustomTextarea
                    name='stack'
                    onChange={this.onChange}
                    state={this.state}
                    label='Стек технологий:'
                />
                <div className="errorMsg">{this.state.errors.stack}</div>
                <CustomTextarea
                    name='project'
                    onChange={this.onChange}
                    state={this.state}
                    label="Описание последнего проекта:"
                />
                <div className="errorMsg">{this.state.errors.project}</div>
                <div className="formButton">
                    <input className="button button--ok" type="submit" value="Сохранить" />
                    <input className="button button--cancel" type="reset" value="Отмена" onClick={this.resetForm} />
                </div>
            </form>
        </>
        );
    }
}
