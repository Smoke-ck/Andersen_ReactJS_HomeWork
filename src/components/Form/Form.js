import React from 'react';
import './Form.css';
import CustomInput from './CustomInput/CustomInput';
import { fetchLogin } from '../../store/actions/loginUser';
import { useState } from 'react'
import { useDispatch } from "react-redux"

function Form({ setActive, callbackStateFromForm }) {

    const ADMIN_NAME = 'mor_2314';
    const ADMIN_PASSWORD = '83r5^_';

    const dispatch = useDispatch();

    const [newForm, setnewForm] = useState({
        name: '',
        password: '',
        errors: {},
    });

    const handleSubmit = (e) => {
        e.preventDefault();
        if (validateForm()) {
            dispatch(fetchLogin(newForm))
            callbackStateFromForm(newForm.name);
            resetForm();
            setActive(false)
        }
    };

    function validateForm() {
        let data = newForm;
        let errors = {};
        let formIsValid = true;
        let { name, password, } = { ...data }

        name = name.trim();
        if (!name) {
            formIsValid = false;
            errors.name = "Поле пустое"
        } else if (name === ADMIN_NAME || name === "Customer") {
            formIsValid = true;
        }
        else if (name !== ADMIN_NAME || name !== "Customer") {
            formIsValid = false;
            errors.name = "Неверный ввод"
        }

        password = password.trim();
        if (!password) {
            formIsValid = false;
            errors.password = "Поле пустое"
        } else if (password === ADMIN_PASSWORD || password === "456") {
            formIsValid = true;
        }
        else if (password !== ADMIN_PASSWORD || password !== "456") {
            formIsValid = false;
            errors.password = "Неверный пароль"
        }
        setnewForm(prevState => ({
            ...prevState,
            errors: errors,
        }));
        return formIsValid;
    }

    const onChange = (e) => {
        setnewForm(prevState => ({
            ...prevState,
            [e.target.name]: e.target.value
        }));
    }

    const resetForm = () => {
        setnewForm({
            name: '',
            password: '',
            errors: '',
        });
    }
    return (<>
        <form className="containerForm" onSubmit={handleSubmit}>
            <p className="formTittle">Login</p>
            <div className="formButton">
                <input className="button button--ok button__login" type="submit" value="Сохранить" />
                <input className="button button--cancel button__login" type="reset" value="Отмена" onClick={resetForm} />
            </div>
            <CustomInput
                type="text"
                name='name'
                onChange={onChange}
                state={newForm}
                label="UserName:"
            />
            <div className="errorMsg">{newForm.errors.name}</div>
            <CustomInput
                type="text"
                name='password'
                onChange={onChange}
                state={newForm}
                label="Pasword:"
            />
            <div className="errorMsg">{newForm.errors.password}</div>
        </form>
    </>
    );
}
export default Form;