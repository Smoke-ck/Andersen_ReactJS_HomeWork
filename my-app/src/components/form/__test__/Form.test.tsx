import Form from "../Form"
import Enzyme, { shallow, render, mount } from "enzyme";

import { Provider } from 'react-redux';
import { shallowToJson } from 'enzyme-to-json';
import configureStore from "redux-mock-store";

 
const mockStore = configureStore([]);
const stringMore = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi porttitor lectus ut sodales porttitor. Mauris ac mattis arcu. Suspendisse id nisl vel lorem efficitur suscipit. Nullam leo quam, ultricies eu porta eget, malesuada at justo. Duis posuere gravida dolor vel tincidunt. In lorem massa, dapibus ut tortor eget, bibendum ultricies velit. Phasellus egestas feugiat augue, at vestibulum tortor sodales quis. Fusce id ante nunc. Praesent aliquam augue nec mauris egestas, at pulvinar turpis ornare. Pellentesque non ex tortor. Aliquam eu sem et lectus tempor auctor et ut metus. Fusce lacinia facilisis ante, at consequat velit elementum id. Suspendisse non turpis interdum, faucibus ligula sed, iaculis purus."
const mockDispatchfn = jest.fn();

describe('Form', () => {
  let wrapper: any;
  let store: any;
  const props: any = {
  createTodo: jest.fn(),
  };
  
  beforeEach(() => {
  store = mockStore({})
    jest.spyOn(store, "dispatch");
    wrapper = mount(
      <Provider store={store}>
        <Form {...props}  dispatch={mockDispatchfn} />
      </Provider>,
    );
    expect(wrapper).toBeDefined();
  });

  it('should match the snapshot', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('should renders form component', () => {
    expect(wrapper.find('[className="form"]').first()).toHaveLength(1);
  });

  it('should component value will be', () => {
    wrapper.find('[className="form__input"]').simulate('change', { target: { value: 'MyTest' } });
    expect(wrapper.find('[className="form__test__text"]').text()).toBe('MyTest');
  });
  it('should renders input field', () => {
    const inputField = wrapper.find('[className="form__input"]');
    expect(inputField.prop('type')).toBe('text');
  });
  it('should renders button', () => {
    const submitButton = wrapper.find('[className="form__button"]');
    expect(submitButton.prop('type')).toBe('submit');
  });
  it("should renders message error", () => {
    const input = wrapper.find('[className="form__input"]');
    input.simulate("change", { target: { value: stringMore } });
    expect(wrapper.find('[className="errorMsg"]').text()).toContain(
      "Превышен лимит текста задачи на"
    );
  });
});




























// describe('Selected component', () => {
//   let wrapper: any;

//   beforeEach(() => {
//     wrapper = mount(<Form{...props} />)
//   })
  
//   afterEach(() => {
//     wrapper.unmount();
//   })

//   it("component should render correctly", () => {
//     expect(wrapper).toBeTruthy();
//   })

//   // it("shoud render popup element", () => {
//   //   const popup = wrapper.find('[className="popup"]');
//   //   expect(popup).toHaveLength(1);
//   // })

//   // it("shoud render 4 button", () => {
//   //   const buttons = wrapper.find('[className="listItem__menu--button"]');
//   //   expect(buttons).toHaveLength(4);
//   // })

//   // it("shoud click for first button", () => {
//   //   wrapper.find("button").at(0).simulate('click');
//   //   expect(props.handleFavoriteToDo).toBeCalled();
//   // })

//   // it("shoud click for second button", () => {

//   //   wrapper.find("button").at(1).simulate('click');
//   //   expect(props.handleDoneToDo).toBeCalled();
//   // })

//   // it("shoud click for tird button", () => {

//   //   wrapper.find("button").at(2).simulate('click');
//   //   expect(props.handleClick).toBeCalled();
//   // })

//   // it("shoud click for fourth button", () => {

//   //   wrapper.find("button").at(3).simulate('click');
//   //   expect(props.handleToggleModal).toBeCalled();
//   // })
// })
