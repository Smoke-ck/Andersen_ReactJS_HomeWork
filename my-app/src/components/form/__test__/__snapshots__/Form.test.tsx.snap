// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Form)
    createTodo={[MockFunction]}
    dispatch={[MockFunction]}
  >
    <Form
      createTodo={[Function]}
      dispatch={[MockFunction]}
    >
      <form
        id="addTaskForm"
        onSubmit={[Function]}
      >
        <div
          className="form"
        >
          <input
            className="form__input"
            id="taskNameInput"
            onChange={[Function]}
            type="text"
            value=""
          />
          <div
            className="two columns"
          >
             
            <input
              className="form__button"
              type="submit"
              value="Добавить"
            />
          </div>
          <p
            className="form__test__text"
          />
        </div>
      </form>
    </Form>
  </Connect(Form)>
</Provider>
`;
